### PR TITLE
YUNIKORN-417: Add "nodes" REST API to fetch list of nodes

### PR DESCRIPTION
**Pull request by manirajv06**
First cut implementation

### CODE REVIEW

File: pkg/webservice/handlers.go [hunk 1]
```diff
@@ -39,6 +39,8 @@ import (
 	"github.com/apache/incubator-yunikorn-core/pkg/scheduler/objects"
 	"github.com/apache/incubator-yunikorn-core/pkg/webservice/dao"
 	"github.com/apache/incubator-yunikorn-scheduler-interface/lib/go/si"
+
+	"github.com/gorilla/mux"
```
Comment (kingamarton) [inline pkg/webservice/handlers.go:43]: This is licensed under BSD 3-Clause "New" or "Revised" License. I am not very familiar with the licensing aspects. Can you check if it is OK to use this? I think we need to mention this at least in some Notice file

**Review by kingamarton — CHANGES_REQUESTED** [review]

File: pkg/webservice/handlers.go (hunk 1 above)
Comment (yangwwei) [reply pkg/webservice/handlers.go:43]: hi @kingamarton  I think this license is fine, and actually, this dependency is already been used by YK. See more info from the dependency doc in https://cwiki.apache.org/confluence/display/INCUBATOR/YuniKornProposal. Close this as resolved.

File: pkg/webservice/handlers.go [hunk 2]
```diff
@@ -521,3 +523,26 @@ func updateConfiguration(conf string) (string, error) {
 	}
 	return "", fmt.Errorf("config plugin not found")
 }
+
+func getPartitionNodes(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	writeHeaders(w)
+	_, partitionExists := vars["partition"]
+	if len(vars) == 0 || !partitionExists {
+		http.Error(w, "Mandatory parameters are missing in URL path. Please check the usage documentation", http.StatusBadRequest)
```
Comment (yangwwei) [inline pkg/webservice/handlers.go:532]: Please explicitly include what parameter is missing in the error message.

File: pkg/webservice/handlers.go [hunk 3]
```diff
@@ -521,3 +523,26 @@ func updateConfiguration(conf string) (string, error) {
...
+		return
+	}
+	partition := schedulerContext.GetPartition(vars["partition"])
```
Comment (yangwwei) [inline pkg/webservice/handlers.go:535]: Can we change line 530 to

```
partitionName, partitionExists := vars["partition"]
```

and here we can reuse the object:
```
partition := schedulerContext.GetPartition(partitionName)
```

**Comment by manirajv06** [comment]
@yangwwei @wilfred-s Made changes to use buildJSONErrorResponse() and added few more test cases. Can you please review?

File: pkg/webservice/handlers.go [hunk 4]
```diff
@@ -610,3 +612,30 @@ func getPartitions(w http.ResponseWriter, r *http.Request) {
...
+	partition, partitionExists := vars["partition"]
+	if !partitionExists {
+		buildJSONErrorResponse(w, "Partition is missing in URL path. Please check the usage documentation", http.StatusBadRequest)
+		return
+	}
+	if len(vars) != 1 {
+		buildJSONErrorResponse(w, "Incorrect URL path. Please check the usage documentation", http.StatusBadRequest)
+		return
+	}
+	partitionContext := schedulerContext.GetPartition(partition)
```
Comment (kingamarton) [inline pkg/webservice/handlers.go:628]: This part is is the same for all the PR's. Can you please extract it in a method and reuse it?

**Review by kingamarton — CHANGES_REQUESTED** [review]
The changes overall looks good to me. I left just on minor comment.

Can you please also rebase it and fix thee conflicts?

**Comment by manirajv06** [comment]
> The changes overall looks good to me. I left just on minor comment.
> 
> Can you please also rebase it and fix thee conflicts?

done

File: pkg/webservice/handlers.go (hunk 4 above)
Comment (manirajv06) [reply pkg/webservice/handlers.go:653]: > This part is is the same for all the PR's. Can you please extract it in a method and reuse it?

Sure, will take this up in a follow up jira once all API changes goes into trunk.

**Review by kingamarton — APPROVED** [review]

**Pull request closed without merging**